File: leetcode/201-300/222-count-complete-tree-nodes/bench/count_nodes.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    long val;
    long left;
    long right;
} Node;

static Node *build(long n) {
    Node *nodes = malloc(n * sizeof(Node));
    for (long i = 0; i < n; i++) {
        long l = 2 * i + 1;
        long r = 2 * i + 2;
        nodes[i].val = i;
        nodes[i].left = l < n ? l : -1;
        nodes[i].right = r < n ? r : -1;
    }
    return nodes;
}

static long left_height(const Node *nodes, long idx) {
    long h = 0;
    long cur = idx;
    while (cur != -1) {
        h++;
        cur = nodes[cur].left;
    }
    return h;
}

static long right_height(const Node *nodes, long idx) {
    long h = 0;
    long cur = idx;
    while (cur != -1) {
        h++;
        cur = nodes[cur].right;
    }
    return h;
}

static long count_nodes(const Node *nodes, long idx) {
    if (idx == -1) return 0;
    long lh = left_height(nodes, idx);
    long rh = right_height(nodes, idx);
    if (lh == rh) {
        return (1L << lh) - 1;
    }
    return 1 + count_nodes(nodes, nodes[idx].left) + count_nodes(nodes, nodes[idx].right);
}

int main(void) {
    long n = 2000000;
    long passes = 2000000;
    long top_span = 2048;
    long modulus = 1000000007;

    Node *nodes = build(n);

    long sink = 0;
    for (long p = 0; p < passes; p++) {
        long start = p % top_span;
        sink = (sink + count_nodes(nodes, start)) % modulus;
    }
    printf("%ld\n", sink);
    free(nodes);
    return 0;
}

File: leetcode/201-300/222-count-complete-tree-nodes/bench/count_nodes.rs
struct Node {
    val: i64,
    left: i64,
    right: i64,
}

fn build(n: i64) -> Vec<Node> {
    let mut nodes: Vec<Node> = Vec::with_capacity(n as usize);
    for i in 0..n {
        let l = 2 * i + 1;
        let r = 2 * i + 2;
        nodes.push(Node {
            val: i,
            left: if l < n { l } else { -1 },
            right: if r < n { r } else { -1 },
        });
    }
    nodes
}

fn left_height(nodes: &[Node], idx: i64) -> i64 {
    let mut h = 0i64;
    let mut cur = idx;
    while cur != -1 {
        h += 1;
        cur = nodes[cur as usize].left;
    }
    h
}

fn right_height(nodes: &[Node], idx: i64) -> i64 {
    let mut h = 0i64;
    let mut cur = idx;
    while cur != -1 {
        h += 1;
        cur = nodes[cur as usize].right;
    }
    h
}

fn count_nodes(nodes: &[Node], idx: i64) -> i64 {
    if idx == -1 {
        return 0;
    }
    let lh = left_height(nodes, idx);
    let rh = right_height(nodes, idx);
    if lh == rh {
        return (1i64 << lh) - 1;
    }
    1 + count_nodes(nodes, nodes[idx as usize].left) + count_nodes(nodes, nodes[idx as usize].right)
}

fn main() {
    let n: i64 = 2000000;
    let passes: i64 = 2000000;
    let top_span: i64 = 2048;
    let modulus: i64 = 1000000007;

    let nodes = build(n);

    let mut sink = 0i64;
    for p in 0..passes {
        let start = p % top_span;
        sink = (sink + count_nodes(&nodes, start)) % modulus;
    }
    println!("{}", sink);
}
